File: ch6/clip5/boj15906/main.cpp
#include <iostream>
#include <queue>

#define MAX_VALUE (1000)

struct point_t
{
    int r;
    int c;
    int changed;
    int cost;
    point_t operator+(point_t rhs)
    {
        return {r + rhs.r, c + rhs.c, changed + rhs.changed, cost + rhs.cost};
    }
    bool operator()(point_t a, point_t b)
    {
        return a.cost > b.cost;
    }
};

point_t direction[4] = {{1, 0, 0, 0}, {0, 1, 0, 0}, {-1, 0, 0, 0}, {0, -1, 0, 0}};

char board[501][501];
int cost[2][501][501];
int main()
{
    int N, t;
    point_t goal;
    scanf("%d %d %d %d", &N, &t, &goal.r, &goal.c);

    for(int r = 1; r <= N; r++)
    {
        for(int c = 1; c <= N; c++)
        {
            scanf(" %c", &board[r][c]);
            cost[0][r][c] = cost[1][r][c] = MAX_VALUE;
        }
    }

    std::priority_queue<point_t, std::vector<point_t>, point_t> pq;

    for(pq.push({1, 1, 0, 0}); ;)
    {
        point_t now = pq.top();
        pq.pop();

        if(now.r == goal.r && now.c == goal.c)
        {
            printf("%d", now.cost);
            break;
        }

        if(cost[now.changed][now.r][now.c] < now.cost)
        {
            continue;
        }

        for(int i = 0; i < 4; i++)
        {
            point_t next = now + direction[i];
            if(next.r <= 0 || next.r > N || next.c <= 0 || next.c > N)
            {
                continue;
            }

            next.changed = 0;
            next.cost = now.cost + 1;

            if(cost[0][next.r][next.c] > next.cost)
            {
                cost[0][next.r][next.c] = next.cost;
                pq.push(next);
            }
        }

        for(int i = 0; i < 4; i++)
        {
            point_t next = now + direction[i];
            if(0 == next.changed)
            {
                next.changed = 1;
                next.cost = now.cost + t;
            }
            next.cost = next.cost + 1;

            for(; ; next = next + direction[i])
            {
                if(next.r <= 0 || next.r > N || next.c <= 0 || next.c > N)
                {
                    break;
                }

                if('#' == board[next.r][next.c])
                {
                    if(cost[1][next.r][next.c] > next.cost)
                    {
                        cost[1][next.r][next.c] = next.cost;
                        pq.push(next);
                    }
                    break;
                }
            }
        }
    }


    return 0;
}

File: ch6/clip5/boj15906/Main.java
import java.util.PriorityQueue;
import java.util.Scanner;

class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int t = sc.nextInt();
        int gr = sc.nextInt();
        int gc = sc.nextInt();
        int dr[] = {-1, 0, 1, 0};
        int dc[] = {0, 1, 0, -1};
        int[][] map = new int[n+ 1][n +1];
        int[][][] cost = new int[2][n + 1][n + 1];

        for(int i = 1; i <= n; i++) {
            String s = sc.next();
            for(int j = 1; j <= n; j++) {
                char ch = s.charAt(j - 1);
                if(ch == '#') map[i][j] = 1;
                else map[i][j] = 0;
                cost[0][i][j] = Integer.MAX_VALUE;
                cost[1][i][j] = Integer.MAX_VALUE;
            }
        }
        PriorityQueue<Point> pq = new PriorityQueue<>(
            (a, b) -> a.cost - b.cost
        );

        pq.add(new Point(1, 1, 0, 0));
        cost[0][1][1] = 0;
        while(!pq.isEmpty()) {
            Point now = pq.poll();
            if(now.r == gr && now.c == gc) {
                System.out.println(now.cost);
                return;
            }
            if(cost[now.changed][now.r][now.c] < now.cost) continue;
            for(int i = 0; i < 4; i++) {
                int nr = now.r + dr[i];
                int nc = now.c + dc[i];
                if(!isValid(nr, nc, n)) continue;
                Point next = new Point(nr, nc, 0, now.cost + 1);
                if(cost[0][nr][nc] > next.cost) {
                    cost[0][nr][nc] = next.cost;
                    pq.add(next);
                }
            }
            for(int i = 0; i < 4; i++) {
                int nr = now.r + dr[i];
                int nc = now.c + dc[i];
                int nextCost = now.cost + t * (now.changed == 0 ? 1 : 0) + 1;
                while(true) {
                    if(!isValid(nr, nc, n)) break;
                    if(map[nr][nc] == 1) {
                        if (cost[1][nr][nc] > nextCost) {
                            cost[1][nr][nc] = nextCost;
                            pq.add(new Point(nr, nc, 1, nextCost));
                        }
                        break;
                    }
                    nr += dr[i];
                    nc += dc[i];
                }
            }
        }
    }
    static boolean isValid(int r, int c, int n) {
        return r >= 1 && r <= n && c >= 1 && c <= n;
    }
}

class Point {
    int r, c, changed, cost;
    Point (int r, int c, int changed, int cost) {
        this.r = r;
        this.c = c;
        this.changed = changed;
        this.cost = cost;
    }
}